### ISSUE TITLE
[followup of #47046] kubelet and btrfs snapshots: failed to get device for dir "/var/lib/kubelet"

### ISSUE THREAD

**Issue by MaximilianMeister** [kubernetes/kubernetes#65204]
> /kind bug

**What happened**:
`/var/lib/kubelet` is a subdirectory of the btrfs subvolume `/var`. kubelet fails to start with the following error:

```
Failed to start ContainerManager failed to get rootfs info: failed to get device for dir "/var/lib/kubelet": could not find device with major: 0, minor: 46 in cached partitions map
```

**What you expected to happen**:
kubelet starts normally and doesn't fail with an error message

**How to reproduce it (as minimally and precisely as possible)**:
* Install a system using btrfs as file system for /
* Create a sub-volume for /var
* Start kubelet (even in disconnected mode)

**Anything else we need to know?**:
This is a follow-up issue of #47046 and has been promoted to fatal from `v1.10` onward, before it was a warning

PR to fix the issue is https://github.com/google/cadvisor/pull/1971

**Environment**:
- Kubernetes version (use `kubectl version`): `1.10` forward
- Cloud provider or hardware configuration: both bare metal and virtual
- OS (e.g. from /etc/os-release): SUSE 12+ or openSUSE (Tumbleweed/Leap)
- Kernel (e.g. `uname -a`): `4.17.1-1`
- Install tools: SUSE/openSUSE rpm
- Others:

**Comment by dashpole** [comment]
fixed by https://github.com/google/cadvisor/pull/1971.
/close

**Comment by cr1st1p** [comment]
If I may - adding some notes - it might be helpful for others, until I get the time to do some more detailed tests and report a bug.

I still had issues with kubelet not starting, with the same error - (version 1.13.4)
My btrfs setup is that btrfs root  is mounted at /, then I have also a subvolume /var and in it a subvolume  /var/lib.
The issue for me I think comes from the fact that I do not have a flat btrfs setup (https://btrfs.wiki.kernel.org/index.php/SysadminGuide#Flat) but a _nested_ one, and as such, I have as a mountpoint ONLY the root directory. 

> Subvolumes don't need to be mounted manually (or via fstab) to their desired locations, they "appear automatically" at their respective locations. 

Manually adding a mount entry for /var (heh - I should probably have mounted /var/lib) in /etc/fstab made it work:
`/dev/mapper/SSD-root`  /var  btrfs rw,relatime,ssd,space_cache,subvolid=259,subvol=/var 0 0`

Also, even without this additional "intermediate" mount point I think it should have been working. I wonder if in cadvisor/.../fs.go, the loop 'for !found {' isn't exiting a tiny bit too fast (without testing for the  root mountpoint itself) . Maybe it should be something like:
```
        for !found {
                pathdir, _ := filepath.Split(dir)
                // trim "/" from the new parent path otherwise the next possible
                // filepath.Split in the loop will not split the string any further
                dir = strings.TrimSuffix(pathdir, "/")
                mount, found = self.mounts[dir]
                // force break when we reach root
                if pathdir == "/" {
                        break
                }
        }

```

**Comment by cr1st1p** [comment]
Has something changed in the code? Kubelet from kubernetes 1.20.1 is failing. 
Doing a mount as I described above fixes it.

**Comment by vorburger** [comment]
see (open) #95826 and (closed) #94335- apparently this was recently fixed (but isn't released yet).
